### PR TITLE
Upgrading Libraries, Scala, SBT and Co-producting Kinesis and NSQ configs 

### PR DESCRIPTION
**Pull request by szareiangm**
Hello Team,

I made the changes below to the codebase:
- Updating Java config library from 1.3.1 to 1.3.3
- Updating Scala scopt library from 3.6.0 to 3.7.0
- Updating Scala pureconfig library from 0.8.0 to 0.9.1
- Updating Scala specs2 library from 3.9.2 to 4.1.0
- Updating Scala scalaz7 library from 7.2.14 to 7.2.22
- Updating Scala snowplowTracker library from 0.3.0 to 0.5.0 and added the ssl configuration for it
- Changed deprecated self-describing Json to SchemaKey
- Updating Scala analyticsSDK library from 0.2.0 to 0.3.0
- Updating Java kinesisClient library from 1.7.5 to 1.9.1 + AWS Kinesis Client Worker was deprecated. - Using Worker.Builder #78
- Separated NSQ and Kinesis as co-products and changed the Hocon file #68
- Updating Scala analyticsSDK library from 0.13.15 to 1.1.5
- Bumping to Scala 2.12.6 from Scala 2.11.11 and using JavaConverters instead of deprecated JavaConversions

CC: @Nafid , @wisdomjiang

### CODE REVIEW

**Comment by snowplowcla** [comment]
@szareiangm has signed the Software Grant and Corporate Contributor License Agreement

File: core/src/main/scala/com.snowplowanalytics.elasticsearch/loader/SnowplowTracking.scala [hunk 1]
```diff
@@ -71,8 +74,8 @@ object SnowplowTracking {
     initialFailureTime: Long,
     message: String): Unit = {
 
-    tracker.trackUnstructEvent(SelfDescribingJson(
-      "iglu:com.snowplowanalytics.monitoring.kinesis/storage_write_failed/jsonschema/1-0-0",
+    tracker.trackSelfDescribingEvent(SelfDescribingData(
+      SchemaKey.fromUri("iglu:com.snowplowanalytics.monitoring.kinesis/storage_write_failed/jsonschema/1-0-0").get,
```
Comment (BenFradet) [inline core/src/main/scala/com.snowplowanalytics.elasticsearch/loader/SnowplowTracking.scala]: If you know the key in advance it's better to use the constructor:
https://github.com/snowplow/iglu-scala-client/blob/f815b2692f1431af7869186b1e9ac8bcaf20f376/src/main/scala/com.snowplowanalytics.iglu/client/SchemaKey.scala#L70-L74

File: core/src/main/scala/com.snowplowanalytics.elasticsearch/loader/ElasticsearchSinkApp.scala [hunk 2]
```diff
@@ -93,21 +95,26 @@ trait ElasticsearchSinkApp {
...
+        props.setProperty(KinesisConnectorConfiguration.PROP_INITIAL_POSITION_IN_STREAM,
+          qc.initialPosition)
+        props.setProperty(KinesisConnectorConfiguration.PROP_MAX_RECORDS,
+          qc.maxRecords.toString)
+        // So that the region of the DynamoDB table is correct
+        props.setProperty(KinesisConnectorConfiguration.PROP_REGION_NAME,
+          qc.region)
+
+
+      case _ => throw new RuntimeException("Kinesis configuration is not valid")
```
Comment (BenFradet) [inline core/src/main/scala/com.snowplowanalytics.elasticsearch/loader/ElasticsearchSinkApp.scala]: I think it's better to have the method return an `Either[String, KinesisConnectorConfiguration]` which we will then combine with other methods which return `Either` in a for comprehension.

For an example (which uses `Validation` but that doesn't matter much) you can look at:

https://github.com/snowplow/snowplow/blob/9b67ff48fbdc6dde0fa11dfd7653a8eed4196647/3-enrich/stream-enrich/pubsub/src/main/scala/com.snowplowanalytics.snowplow.enrich.stream/sources/GooglePubSubSource.scala#L48-L68

**Comment by szareiangm** [comment]
@BenFradet I rebased the commits.  will address the comments tomorrow.

**Comment by szareiangm** [comment]
comments addressed. Good to go @BenFradet

File: core/src/main/scala/com.snowplowanalytics.elasticsearch/loader/ElasticsearchSinkApp.scala [hunk 3]
```diff
@@ -144,33 +154,43 @@ trait ElasticsearchSinkApp {
 
     val badSink = conf.sink.bad match {
       case "stderr" => new StdouterrSink
-      case "nsq" => new NsqSink(conf)
+      case "nsq" =>
+        conf.queueConfig match {
+          case  qc: Nsq =>
+            new NsqSink(qc.host, qc.port, conf.streams.outStreamName)
+          case _ => throw new RuntimeException("queueConfig is not valid for Nsq")
```
Comment (BenFradet) [inline core/src/main/scala/com.snowplowanalytics.elasticsearch/loader/ElasticsearchSinkApp.scala]: same remark here

File: core/src/main/scala/com.snowplowanalytics.elasticsearch/loader/ElasticsearchSinkApp.scala [hunk 4]
```diff
@@ -144,33 +154,43 @@ trait ElasticsearchSinkApp {
...
       case "kinesis" =>
-        val kinesisSinkName = conf.streams.outStreamName
-        val kinesisSinkRegion = conf.kinesis.region
-        val kinesisSinkEndpoint = conf.kinesis.endpoint
-        new KinesisSink(credentials, kinesisSinkEndpoint, kinesisSinkRegion, kinesisSinkName)
-    }
+        conf.queueConfig match {
+          case  qc: Kinesis =>
+            new KinesisSink(credentials, qc.endpoint, qc.region, conf.streams.outStreamName)
+          case _ => throw new RuntimeException("queueConfig is not valid for Kinesis")
```
Comment (BenFradet) [inline core/src/main/scala/com.snowplowanalytics.elasticsearch/loader/ElasticsearchSinkApp.scala]: and here

File: core/src/main/scala/com.snowplowanalytics.elasticsearch/loader/NsqSourceExecutor.scala [hunk 5]
```diff
@@ -107,18 +104,23 @@ class NsqSourceExecutor(
...
+      case qc: Nsq =>
+        lookup.addLookupAddress(qc.host, qc.lookupPort)
+        val consumer = new NSQConsumer(lookup,
+          config.streams.inStreamName,
+          qc.channelName,
+          nsqCallback,
+          new NSQConfig(),
+          errorCallback)
+        consumer.start()
+      case _ => throw new RuntimeException("queueConfig is not valid for Nsq")
```
Comment (BenFradet) [inline core/src/main/scala/com.snowplowanalytics.elasticsearch/loader/NsqSourceExecutor.scala]: and here

File: examples/config.hocon.sample [hunk 6]
```diff
@@ -55,8 +55,10 @@ aws {
   secretKey = iam
 }
 
-# config for NSQ
-nsq {
+queueConfig {
```
Comment (BenFradet) [inline examples/config.hocon.sample]: We use `source` in stream enrich, so I think we should try to keep things consistent here too

Comment (BenFradet) [reply examples/config.hocon.sample]: ah, just saw your remark in #68, we went with `sourceSink` in stream enrich

File: core/src/main/scala/com.snowplowanalytics.elasticsearch/loader/NsqSourceExecutor.scala [hunk 7]
```diff
@@ -106,19 +103,24 @@ class NsqSourceExecutor(
...
+        lookup.addLookupAddress(queue.host, queue.lookupPort)
+        val consumer = new NSQConsumer(lookup,
+          config.streams.inStreamName,
+          queue.channelName,
+          nsqCallback,
+          new NSQConfig(),
+          errorCallback)
+        consumer.start()
+      case _ => throw new IllegalArgumentException("queue config is not valid for Nsq")
+    }
```
Comment (BenFradet) [inline core/src/main/scala/com.snowplowanalytics.elasticsearch/loader/NsqSourceExecutor.scala]: this has already been validated so you should be able to directly pass a queue config which has type `Nsq`

File: http/src/main/scala/com.snowplowanalytics.snowplow.storage.kinesis.elasticsearch/clients/SignedHttpClientConfigCallback.scala [hunk 8]
```diff
@@ -52,7 +52,7 @@ class SignedHttpClientConfigCallback(
 ) extends HttpClientConfigCallback {
   private def clock(): LocalDateTime = LocalDateTime.now(ZoneId.of("UTC"))
   private val service = "es"
-  private val signer = AwsSigner(credentialsProvider, region, service, clock)
+  private val signer = AwsSigner(credentialsProvider, region, service, () => SignedHttpClientConfigCallback.this.clock())
```
Comment (BenFradet) [inline http/src/main/scala/com.snowplowanalytics.snowplow.storage.kinesis.elasticsearch/clients/SignedHttpClientConfigCallback.scala]: why was this necessary?

File: core/src/main/scala/com.snowplowanalytics.elasticsearch/loader/ElasticsearchSinkApp.scala [hunk 9]
```diff
@@ -126,7 +132,10 @@ trait ElasticsearchSinkApp {
     props.setProperty(KinesisConnectorConfiguration.PROP_CONNECTOR_DESTINATION, "elasticsearch")
     props.setProperty(KinesisConnectorConfiguration.PROP_RETRY_LIMIT, "1")
 
-    new KinesisConnectorConfiguration(props, CredentialsLookup.getCredentialsProvider(config.aws.accessKey, config.aws.secretKey))
+    new KinesisConnectorConfiguration(props,
+      CredentialsLookup.getCredentialsProvider(config.aws.accessKey,
+        config.aws.secretKey)
+    ).success
```
Comment (BenFradet) [inline core/src/main/scala/com.snowplowanalytics.elasticsearch/loader/ElasticsearchSinkApp.scala]: here you're discarding the result of https://github.com/snowplow/snowplow-elasticsearch-loader/pull/81/files#diff-6a74f84ccb19a31030747933ed804b27R98 and as a result, this function will always return a success.

Whatt you can do is just put every `setPropery` in the `case queue: Kinesis` and build your `KinesisConnectorConfiguration` in the case directly.

Does that make sense?

File: http/src/main/scala/com.snowplowanalytics.snowplow.storage.kinesis.elasticsearch/clients/SignedHttpClientConfigCallback.scala (hunk 8 above)
Comment (szareiangm) [reply http/src/main/scala/com.snowplowanalytics.snowplow.storage.kinesis.elasticsearch/clients/SignedHttpClientConfigCallback.scala]: to mute this sbt compile warning:
`[warn] snowplow-elasticsearch-loader/http/src/main/scala/com.snowplowanalytics.snowplow.storage.kinesis.elasticsearch/clients/SignedHttpClientConfigCallback.scala:55:72: Eta-expansion of zero-argument methods is deprecated. To avoid this warning, write (() => SignedHttpClientConfigCallback.this.clock()).
[warn]   private val signer = AwsSigner(credentialsProvider, region, service, clock)`

**Comment by szareiangm** [comment]
I addressed the comments, I think it is in good shape for now.
